File: P2P-cpp/GraphManager.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <thread>
#include <queue>
#include <zmq.hpp>
#include <arpa/inet.h>
#include <netdb.h>
#include "Graph.cpp"  // Graphクラスのヘッダーファイル
#include "Message.cpp" // Messageクラスのヘッダーファイル

class GraphManager {
public:
    GraphManager(const std::string& id, Graph* graph, const std::string& ip_addr)
        : id(id), graph(graph), ip_addr(ip_addr), port(10010) {
        start();
    }

    static GraphManager* initForEspresso(const std::string& dir_path) {
        char hostname[1024];
        gethostname(hostname, sizeof(hostname));
        std::string host_name = hostname;
        std::string host_ip = getHostIP(host_name);
        std::string graph_txt_file = dir_path + host_name + ".txt";

        std::ifstream f(graph_txt_file);
        std::string line;
        std::unordered_map<std::string, Node*> nodes;
        std::unordered_map<std::string, std::vector<std::string>> ADJ;

        while (std::getline(f, line)) {
            std::stringstream ss(line);
            std::string edge[3];
            for (int i = 0; i < 3; ++i) {
                std::getline(ss, edge[i], ',');
            }
            std::string src = edge[0], dest = edge[1], dest_ip = edge[2];

            if (nodes.find(src) == nodes.end()) {
                nodes[src] = new Node(src, host_ip);
            }
            if (ADJ.find(src) == ADJ.end()) {
                ADJ[src] = std::vector<std::string>();
            }
            if (nodes.find(dest) == nodes.end()) {
                nodes[dest] = new Node(dest, dest_ip);
            }
            ADJ[src].push_back(dest);
        }

        f.close();

        Graph* graph = new Graph(ADJ, nodes);
        return new GraphManager(host_name, graph, host_ip);
    }

    void start() {
        std::thread thr_random_walk(&GraphManager::random_walk, this);
        thr_random_walk.detach();
        std::thread thr_notify_result(&GraphManager::notify_result, this);
        thr_notify_result.detach();
        std::thread thr_send_message(&GraphManager::send_message, this);
        thr_send_message.detach();
        std::thread thr_receive_message(&GraphManager::receive_message, this);
        thr_receive_message.detach();
        std::cout << "GraphManager started. IP addr: " << ip_addr << ", recv port: " << port << std::endl;
    }

private:
    std::string id;
    Graph* graph;
    std::string ip_addr;
    int port;
    std::queue<Message> receive_queue;
    std::queue<Message> send_queue;
    std::queue<std::pair<std::string, std::unordered_map<std::string, int>>> notify_queue;

    void random_walk() {
        while (true) {
            Message message = receive_queue.front();
            receive_queue.pop();

            // notify immediately when source is dangling node.
            if (graph->getNodes().find(message.source_id) == graph->getNodes().end()) {
                notify_queue.push({ message.user, {{message.source_id, message.count}} });
                continue;
            }
            auto end_walk = graph->random_walk(message.source_id, message.count, message.alpha);
            if (!end_walk.empty()) {
                notify_queue.push({ message.user, end_walk });
            }
            // Handle escaped walk if needed
        }
    }

    void notify_result() {
        zmq::context_t context(1);
        zmq::socket_t socket(context, ZMQ_PUSH);
        while (true) {
            auto notify = notify_queue.front();
            notify_queue.pop();
            socket.connect("tcp://" + notify.first + ":" + std::to_string(port));
            socket.send(zmq::buffer(notify.second), zmq::send_flags::none);
            std::cout << "Notified to " << notify.first << std::endl;
        }
    }

    void send_message() {
        zmq::context_t context(1);
        zmq::socket_t socket(context, ZMQ_PUSH);
        while (true) {
            Message message = send_queue.front();
            send_queue.pop();
            socket.connect("tcp://" + message.GM + ":" + std::to_string(port));
            socket.send(zmq::buffer(message), zmq::send_flags::none);
            std::cout << "Sent to " << message.GM << std::endl;
        }
    }

    void receive_message() {
        zmq::context_t context(1);
        zmq::socket_t socket(context, ZMQ_PULL);
        socket.bind("tcp://" + ip_addr + ":" + std::to_string(port));
        while (true) {
            zmq::message_t message;
            socket.recv(message);
            Message msg = Message::from_bytes(message.data());
            receive_queue.push(msg);
            std::cout << "Received message\nsource " << msg.source_id << ", count " << msg.count << std::endl;
        }
    }

    static std::string getHostIP(const std::string& host_name) {
        struct hostent* he = gethostbyname(host_name.c_str());
        if (he == nullptr) {
            perror("gethostbyname");
            return "";
        }
        struct in_addr** addr_list = (struct in_addr**)he->h_addr_list;
        return inet_ntoa(*addr_list[0]);
    }
};

int main(int argc, char* argv[]) {
    if (argc != 2) {
        std::cerr << "Usage: " << argv[0] << " <directory_path>" << std::endl;
        return 1;
    }

    GraphManager* gm = GraphManager::initForEspresso(argv[1]);
    // Do necessary cleanup and exit
    delete gm;
    return 0;
}

File: simple/default/GraphManager.py
from Graph import *
from Message import *
import threading
import os
import socket
import sys
import json
import zmq


class GraphManager:
    def __init__(self, id, graph, ip_addr):
        self.id = id
        self.graph = graph
        self.ip_addr = ip_addr
        self.port = 10026
        self.receive_queue = Queue()
        self.send_queue = Queue()
        self.notify_queue = Queue()
        self.across_server_count_total = 0
        self.start()
        # self.total_move_count = 0

    @classmethod
    def init_for_espresso(cls, dir_path):
        host_name = os.uname()[1]
        host_ip = socket.gethostbyname(host_name)
        graph_txt_file = dir_path + host_name + ".txt"
        f = open(graph_txt_file)
        data = f.read()
        f.close()
        data = data.split("\n")
        del data[-1]
        for i in range(len(data)):
            data[i] = data[i].split(",")

        ADJ = dict()
        nodes = dict()
        for line in data:
            edge = [line[0], line[1]]
            dest_ip = line[2]
            if edge[0] not in nodes.keys():
                nodes[edge[0]] = Node(edge[0], host_ip)
            if edge[0] not in ADJ.keys():
                ADJ[edge[0]] = list()
            if edge[1] not in nodes.keys():
                nodes[edge[1]] = Node(edge[1], dest_ip)
            ADJ[edge[0]].append(edge[1])

        graph = Graph(ADJ, nodes)
        gm = GraphManager(host_name, graph, host_ip)
        gm.host_name = host_name

        return gm

    def __repr__(self):
        rtn = ""
        rtn += "Graph Manager {}\n".format(self.id)
        for node in self.graph.nodes.values():
            rtn += "Node: {}, ADJ:".format(node.id)
            for adj_node in node.adj.values():
                rtn += " {}({}),".format(adj_node.id, adj_node.manager)
            rtn = rtn[:-1] + "\n"
        return rtn

    def start(self):
        thr_random_walk = threading.Thread(target=self.random_walk, daemon=False)
        thr_random_walk.start()
        thr_notify_result = threading.Thread(target=self.notify_result, daemon=False)
        thr_notify_result.start()
        thr_send_message = threading.Thread(target=self.send_message, daemon=False)
        thr_send_message.start()
        thr_receive_message = threading.Thread(
            target=self.receive_message, daemon=False
        )
        thr_receive_message.start()
        print(
            "GraphManager started. IP addr: {}, recv port: {}".format(
                self.ip_addr, self.port
            )
        )

    """
    メッセージを取り出す
    処理を実行
    処理の結果、それを再びキューに格納するのかを指定
    """

    def random_walk(self):
        # キューからメッセージがなくなるまで繰り返す
        while True:
            print("Random Walk")
            message = self.receive_queue.get()
            print("Processing Message: ", message)
            # print('Processing Message: ', message)
            # つまり、ここでRWではなく、確立Nで終わったか次のサーバに移動するのかのいずれかを決定する
            end_walk, escaped_walk = self.graph.random_walk(
                message.source_id, message.count, message.alpha
            )
            print("end_walk: {}, escaped_walk: {}".format(end_walk, escaped_walk))
            # # print('end_walk: {}, escaped_walk: {}'.format(end_walk, escaped_walk))
            # RWが終了したときの処理
            if len(end_walk) > 0:
                self.notify_queue.put([message.user, end_walk])
            # RWが継続するときの処理
            if len(escaped_walk) > 0:
                self.across_server_count_total += 1
                for node_id, val in escaped_walk.items():
                    print("Escaped Walk: kokokokokokoko", node_id, val)
                    self.send_queue.put(
                        Message(
                            node_id,
                            val,
                            self.graph.outside_nodes[node_id].manager,
                            message.user,
                            message.alpha,
                        )
                    )

    def notify_result(self):
        while True:
            user, end_walk = self.notify_queue.get()
            context = zmq.Context()
            socket = context.socket(zmq.PUSH)
            socket.connect("tcp://{}:{}".format(user, self.port))
            socket.send(str(end_walk).encode("utf-8"))
            socket.close()
            context.destroy()
            print("Notified to {}\n{}".format(user, end_walk))

    def send_message(self):
        while True:
            message = self.send_queue.get()
            context = zmq.Context()
            socket = context.socket(zmq.PUSH)
            socket.connect("tcp://{}:{}".format(message.GM, self.port))
            socket.send(bytes(message))
            socket.close()
            context.destroy()
            print("Sent to {}\n{}".format(message.GM, message))

    def receive_message(self):
        # ここの回数ー初めに命令サーバから出された時
        context = zmq.Context()
        socket = context.socket(zmq.PULL)
        socket.bind("tcp://{}:{}".format(self.ip_addr, self.port))
        while True:
            message_bytes = socket.recv()
            message = Message.from_bytes(message_bytes)
            self.receive_queue.put(message)
            print(
                "Recieved message\nsource {}, count {}".format(
                    message.source_id, message.count
                )
            )
            # self.total_move_count += 0


if __name__ == "__main__":
    gm = GraphManager.init_for_espresso(sys.argv[1])
    try:
        while True:
            pass
    except KeyboardInterrupt:
        # 終了時に 独自のフォルダにログを保存
        print("Exiting...")
        print("Total moves across servers: {}".format(gm.across_server_count_total))
        if not os.path.exists("./logs"):
            os.mkdir("./logs")
        with open("./logs/count.txt".format(gm.host_name), "w") as f:
            f.write(
                "Total moves across servers: {}\n".format(gm.across_server_count_total)
            )